### PR TITLE
Tom end to end

### PR DESCRIPTION
**Pull request by teh**
I know the type stuff is getting a bit WTF but I think I managed to contain most of the weird type errors in Resolver.hs by avoiding non-closed transforms on types the users have to implement. I probably missed some, remains to be seen.

Tests are failing and the enum implementation itself is still buggy (need to figure out how to extract the name from a `ValueEnum`). Will ping this branch when fixed.

### CODE REVIEW

**Comment by jml** [comment]
Thanks!

I've incorporated some of the fixes into my own branch. I'm going to hold off on the renames for now—not because I think they are necessarily a bad idea, but because I don't understand my own examples well enough yet to pick good names, hence the deliberately awful `ServerDog`.

Some questions to follow soon.

File: tests/ValueTests.hs [hunk 1]
```diff
@@ -52,9 +53,9 @@ tests = testSpec "Value" $ do
     prop "Bool" $ prop_roundtripValue @Bool
     prop "Int32" $ prop_roundtripValue @Int32
     prop "Double" $ prop_roundtripValue @Double
-    prop "Text" $ forAll arbitraryText prop_roundtripValue
+    prop "Text" $ forAll arbitraryText (prop_roundtripValue @Text)
     prop "Lists" $ prop_roundtripValue @[Int32]
-    prop "Non-empty lists" $ forAll (arbitraryNonEmpty @Int32) prop_roundtripValue
+    prop "Non-empty lists" $ forAll (arbitraryNonEmpty @Int32) (prop_roundtripValue @(NonEmpty Int32))
```
Comment (jml) [inline tests/ValueTests.hs]: Are these changes necessary?

Comment (teh) [inline tests/ValueTests.hs]: Ah no, leftover from when I was trying the associated type family for `FromValue`, removed

**Comment by jml** [comment]
I'll incorporate this into my branch with a little rebasing, if that's OK with you.

**Comment by teh** [comment]
Please do!

**Comment by jml** [comment]
Closing since #77 is merged and most of these changes were included in it.
